### PR TITLE
Add 3D attention mask to T5 model (#9643)

### LINKED ISSUE #9643: [Feature Request] Add 3D attention mask for T5Model
## Environment info

- `transformers` version: 4.2.1
- Platform: Linux
- Python version: 3.6.9
- PyTorch version (GPU?): 1.7.0+cuda11.0

### Who can help
T5: @patrickvonplaten

## Information
The `get_extended_attention_mask()` does not sufficiently address the case with 3D attention mask. The problem emerges for T5Model as `input_ids` and `decoder_input_ids` are of different length and the `attention_mask` is of shape [Batch_size, Seq_length, Seq_length]. The decoder uses `attention_mask` directly for `encoder_attention_mask` in cross-attention, which is of incorrect shape and the error message does give any information about why it happens. 

## To reproduce
As described above. I can add code later if needed. 

## Expected behavior
I propose to add a sanity check for attention masks or improve the `get_extended_attention_mask()` method.

### PR DESCRIPTION
**Pull request by lexhuismans**
# What does this PR do?
It allows for 3D attention mask in T5 model (modeling_t5.py). 

Fixes #9643 

This  is a solution for allowing the 3D attention mask in the T5 model by making it broadcastable. It is based on what is used in BERT.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @lexhuismans,

Thanks a lot for your PR! Could you also add a test to verify that T5 can be used with a 3D mask?

**Comment by lexhuismans** [comment]
Hey @patrickvonplaten,

Thanks for your message. I had a shot in adding a test for the 3D attention mask. The test passed on my device. 

I based the test on a similar test for a default attention mask in test_modeling_bert.py. (Not sure if bert already tests for 3D attention mask?) 

Also, I did a rebase before pushing which is why there are so many other commits in-between. 

Let me know if something is still missing or incorrect so I can have a look at it.

**Comment by lexhuismans** [comment]
I made a new PR with just the two commits #11197.
